crates/ra_hir/src/resolve.rs

```diff
@@ -0,0 +1,413 @@
+                    (resolved, reached_fixed_point)
+                } else {
+                    map_resolution(resolve_path_segments(
+                        db,
+                        krate_item_map,
+                        resolved.and_then(|res| match res {
+                            Resolution::Def { def } => Some(def),
+                            _ => None,
+                        }),
+                        &path.segments[1..],
+                    ))
+                }
+            }
+            PathKind::Super => {
+                let (item_map, original_module) = match self.module() {
+                    Some(m) => m,
+                    _ => return (PerNs::none(), ReachedFixedPoint::Yes),
+                };
+                let parent = match original_module.parent(db) {
+                    Some(p) => p,
+                    _ => {
+                        log::debug!("super path in root module");
+                        return (PerNs::none(), ReachedFixedPoint::Yes);
+                    }
+                };
+                map_resolution(resolve_path_segments(
+                    db,
+                    Some((original_module.krate, item_map)),
+                    PerNs::types(parent.into()),
+                    &path.segments,
+                ))
+            }
+            PathKind::Abs => {
+                // TODO: absolute use is not supported
+                return (PerNs::none(), ReachedFixedPoint::Yes);
+            }
+        }
+    }
+
+    pub fn all_names(&self) -> FxHashMap<Name, PerNs<Resolution>> {
+        let mut names = FxHashMap::default();
+        for scope in &self.scopes {
+            scope.collect_names(&mut |name, res| {
+                let current: &mut PerNs<Resolution> = names.entry(name).or_default();
+                if current.types.is_none() {
+                    current.types = res.types;
+                }
+                if current.values.is_none() {
+                    current.values = res.values;
+                }
+            });
+        }
+        names
+    }
+
+    fn module(&self) -> Option<(&ItemMap, Module)> {
+        for scope in self.scopes.iter().rev() {
+            match scope {
+                Scope::ModuleScope(m) => {
+                    return Some((&m.item_map, m.module.clone()));
+                }
+                Scope::ModuleScopeRef(m) => {
+                    return Some((m.item_map, m.module.clone()));
+                }
+                _ => {}
+            }
+        }
+        None
+    }
+}
+
+fn map_resolution(
+    res: (PerNs<ModuleDef>, ReachedFixedPoint),
+) -> (PerNs<Resolution>, ReachedFixedPoint) {
+    (res.0.map(|def| Resolution::Def { def }), res.1)
+}
+
+fn resolve_path_segments(
+    db: &impl HirDatabase,
+    // during import resolution for a certain, we can't use the query to get the
+    // item map for that crate since it's just being constructed. So in that
+    // case, we need to pass the item map here.
```

Comment:
Yeah, this is this awkward bit of “an abstraction interface has two sides: “public”, for consumers, and “priviate”, for implemenetrs. It sucks when one is round and he other is squared”